hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterRpcClient.java

```diff
@@ -471,6 +471,9 @@ private Object invokeMethod(
         if (this.rpcMonitor != null) {
           this.rpcMonitor.proxyOpComplete(true);
         }
+        if (this.router.getRouterMetrics() != null) {
+          this.router.getRouterMetrics().incInvokedMethod(method);
```

Comment:
Can you point me to the equivalent for the Namenode?
I thought that the RouterRpcServer was already tracking most of these metrics.

Reply:
@goiri Thanks for the review.
A similar metrics for NameNode is "org.apache.hadoop.hdfs.server.namenode.metrics.NameNodeMetrics".
I think this RouterMetrics is trying to monitor from a different view of RouterRpcServer. IMHO, RouterRpcServer is monitoring Router as a Server, the new metrics would be treating Router as a Client. Some operations are not triggered by user requests but from Router itself.
In our cluster, Router's async call queue was jammed, but we couldn't get the operations from the RouterRpcServer metrics. Then we found most of the operations are "getDatanodeReport" which are invoked by Router's FederationMetrics and not by users.

Reply:
That makes sense.
I have never been able to find the code that the RpcServer uses to track each method; does it align with this?
We may want to put these methods within a more intuitive namespace referring to the client too.

Reply:
I think it's the `metrics.incrGetBlockLocations` in _NameNodeRpcServer_.
```
  @Override // ClientProtocol
  public LocatedBlocks getBlockLocations(String src, long offset, long length) 
      throws IOException {
    checkNNStartup();
    metrics.incrGetBlockLocations();
    LocatedBlocks locatedBlocks =
        namesystem.getBlockLocations(getClientMachine(), src, offset, length);
    return locatedBlocks;
```
A similar way to record metrics would be to add the increment method in `RouterClientProtocol` I think. But for Routers, there is a little difference, take `RBFMetrics.getNodeUsage` for example, it will not call `RouterClientProtocol.getDatanodeReport`, but directly invoke the method with Reflection in `RouterRpcServer`. Since Reflection is widely used in Routers to connect to NameNode, IMHO, it's also convenient to track the metrics?

Reply:
> We may want to put these methods within a more intuitive namespace referring to the client too.
Could you be more specific?

Reply:
You are adding all these metrics raw to RouterMetrics.
I'm wondering if we should have something that refers to this being metrics for the client.

Reply:
The metrics name is "RouterActivity", seems a little related. Or do we need to create a new one named "RouterClientActivity"?

Reply:
I think it would be better.